File: kse-frontend/src/main/java/kse/frontend/cli/config/ErrorDecoder.kt
package kse.frontend.cli.config

import com.github.yingzhuo.carnival.exception.business.BusinessException
import com.github.yingzhuo.carnival.openfeign.error.AbstractErrorDecoder
import com.github.yingzhuo.carnival.spring.JacksonUtils
import org.springframework.stereotype.Component
import java.io.Reader
import java.util.*

@Component
object ErrorDecoder : AbstractErrorDecoder(), feign.codec.ErrorDecoder {

    override fun decode(methodKey: String, status: Int, body: Reader, headers: MutableMap<String, MutableCollection<String>>): Optional<Exception> {
        if (status == 400) {
            val info = JacksonUtils.readValue(body, Map::class.java)
            if (info["type"] == BusinessException::class.java.name) {
                return Optional.of(
                        BusinessException(
                                info["code"] as String,
                                info["message"] as String
                        )
                )
            }
        }
        return Optional.empty()
    }

}

File: kse-frontend/src/main/java/kse/frontend/cli/config/ErrorDecoder.scala
package kse.frontend.cli.config

import java.io.Reader
import java.util
import java.util.Optional

import com.github.yingzhuo.carnival.exception.business.BusinessException
import com.github.yingzhuo.carnival.openfeign.error.AbstractErrorDecoder
import org.apache.commons.io.IOUtils
import org.springframework.stereotype.Component

@Component
protected class ErrorDecoder extends AbstractErrorDecoder {

  override def decode(methodKey: String, status: Int, body: Reader, headers: util.Map[String, util.Collection[String]]): Optional[Exception] = {

    if (status != 503) {
      return Optional.empty()
    }

    IOUtils.toString(body) match {
      case ErrorDecoder.ExceptionPattern(code, message, exceptionType) =>
        Optional.of(newException(code, message, exceptionType))
      case _ =>
        Optional.empty();
    }
  }

  private def newException(code: String, message: String, exceptionType: String): Exception =
    exceptionType match {
      case _ if exceptionType == classOf[BusinessException].getName =>
        BusinessException.of(code, message)
      case _ =>
        null
    }

}


protected object ErrorDecoder {
  private val ExceptionPattern = "@@(.+)@@(.+)@@(.+)".r
}
